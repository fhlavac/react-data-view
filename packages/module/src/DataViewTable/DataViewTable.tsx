import React, { ReactNode } from 'react';
import {
  Table,
  TableProps,
  Tbody,
  Td,
  TdProps,
  Th,
  Thead,
  ThProps,
  Tr,
  TrProps
} from '@patternfly/react-table';
import { useInternalContext } from '../InternalContext';

export type DataViewTh = ReactNode | { cell: ReactNode; props?: ThProps };
export type DataViewTd = ReactNode | { cell: ReactNode; props?: TdProps };
export type DataViewTr = DataViewTd[] | { row: DataViewTd[], id?: string, props?: TrProps };

export const isDataViewThObject = (value: DataViewTh): value is { cell: ReactNode; props?: ThProps } => value != null && typeof value === 'object' && 'cell' in value;
export const isDataViewTdObject = (value: DataViewTd): value is { cell: ReactNode; props?: TdProps } => value != null && typeof value === 'object' && 'cell' in value;
export const isDataViewTrObject = (value: DataViewTr): value is { row: DataViewTd[], id?: string } => value != null && typeof value === 'object' && 'row' in value;


export interface DataViewTableProps extends Omit<TableProps, 'onSelect' | 'rows'> {
  /** Columns definition */
  columns: DataViewTh[];
  /** Current page rows */
  rows: DataViewTr[];
  /** Custom OUIA ID */
  ouiaId?: string;
}

export const DataViewTable: React.FC<DataViewTableProps> = ({
  columns,
  rows,
  ouiaId = 'DataViewTable',
  ...props
}: DataViewTableProps) => {
  const { selection } = useInternalContext();
  const { onSelect, isSelected, isSelectDisabled } = selection ?? {};

  return (
    <Table aria-label="Data table" ouiaId={ouiaId} {...props}>
      <Thead data-ouia-component-id={`${ouiaId}-thead`}>
        <Tr ouiaId={`${ouiaId}-tr-head`}>
          {onSelect && isSelected && <Th key="row-select" />}
          {columns.map((column, index) => (
            <Th
              key={index}
              {...(isDataViewThObject(column) && (column?.props ?? {}))}
              data-ouia-component-id={`${ouiaId}-th-${index}`}
            >
              {isDataViewThObject(column) ? column.cell : column}
            </Th>
          ))}
        </Tr>
      </Thead>
      <Tbody>
        {rows.map((row, rowIndex) => { 
          const rowIsObject = isDataViewTrObject(row);
          return (
            <Tr key={rowIndex} ouiaId={`${ouiaId}-tr-${rowIndex}`} {...(rowIsObject && (row?.props ?? {}))}>
              {onSelect && isSelected && (
                <Td
                  key={`select-${rowIndex}`}
                  select={{
                    rowIndex,
                    onSelect: (_event, isSelecting) => {
                      onSelect?.(isSelecting, rowIsObject ? row : [ row ])
                    },
                    isSelected: isSelected?.(row) || false,
                    isDisabled: isSelectDisabled?.(row) || false,
                  }}
                />
              )}
              {(rowIsObject ? row.row : row).map((cell, colIndex) => {
                const cellIsObject = isDataViewTdObject(cell);
                return (
                  <Td
                    key={colIndex}
                    {...(cellIsObject && (cell?.props ?? {}))}
                    data-ouia-component-id={`${ouiaId}-td-${rowIndex}-${colIndex}`}
                  >
                    {cellIsObject ? cell.cell : cell}
                  </Td>
                )
              })}
            </Tr>
          )})}
      </Tbody>
    </Table>
  );
};

export default DataViewTable;
