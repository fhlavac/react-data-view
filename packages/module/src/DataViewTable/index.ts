export { default } from './DataViewTable';
export * from './DataViewTable';
