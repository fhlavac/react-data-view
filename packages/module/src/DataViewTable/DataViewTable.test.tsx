import React from 'react';
import { render } from '@testing-library/react';
import { DataViewTable } from '@patternfly/react-data-view/dist/dynamic/DataViewTable';

interface Repository {
  name: string;
  branches: string | null;
  prs: string | null;
  workspaces: string;
  lastCommit: string;
}

const repositories: Repository[] = [
  { name: 'one', branches: 'two', prs: 'three', workspaces: 'four', lastCommit: 'five' },
  { name: 'one - 2', branches: null, prs: null, workspaces: 'four - 2', lastCommit: 'five - 2' },
  { name: 'one - 3', branches: 'two - 3', prs: 'three - 3', workspaces: 'four - 3', lastCommit: 'five - 3' },
  { name: 'one - 4', branches: 'two - 4', prs: 'null', workspaces: 'four - 4', lastCommit: 'five - 4' },
  { name: 'one - 5', branches: 'two - 5', prs: 'three - 5', workspaces: 'four - 5', lastCommit: 'five - 5' },
  { name: 'one - 6', branches: 'two - 6', prs: 'three - 6', workspaces: 'four - 6', lastCommit: 'five - 6' }
];

const rows = repositories.map(repo => ({
  row: Object.values(repo),
}));

const columns = [ 'Repositories', 'Branches', 'Pull requests', 'Workspaces', 'Last commit' ];

const ouiaId = 'TableExample';

describe('DataViewTable component', () => {
  test('should render correctly', () => {
    const { container } = render(
      <DataViewTable aria-label='Repositories table' ouiaId={ouiaId} columns={columns} rows={rows} />
    );
    expect(container).toMatchSnapshot();
  });
});
