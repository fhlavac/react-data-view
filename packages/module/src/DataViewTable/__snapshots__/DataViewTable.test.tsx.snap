// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataViewTable component should render correctly 1`] = `
<div>
  <table
    aria-label="Repositories table"
    class="pf-v5-c-table pf-m-grid-md"
    data-ouia-component-id="TableExample"
    data-ouia-component-type="PF5/Table"
    data-ouia-safe="true"
    role="grid"
  >
    <thead
      class="pf-v5-c-table__thead"
      data-ouia-component-id="TableExample-thead"
    >
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="TableExample-tr-head"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="TableExample-th-0"
          scope="col"
          tabindex="-1"
        >
          Repositories
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="TableExample-th-1"
          scope="col"
          tabindex="-1"
        >
          Branches
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="TableExample-th-2"
          scope="col"
          tabindex="-1"
        >
          Pull requests
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="TableExample-th-3"
          scope="col"
          tabindex="-1"
        >
          Workspaces
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="TableExample-th-4"
          scope="col"
          tabindex="-1"
        >
          Last commit
        </th>
      </tr>
    </thead>
    <tbody
      class="pf-v5-c-table__tbody"
      role="rowgroup"
    >
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="TableExample-tr-0"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-0-0"
          tabindex="-1"
        >
          one
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-0-1"
          tabindex="-1"
        >
          two
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-0-2"
          tabindex="-1"
        >
          three
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-0-3"
          tabindex="-1"
        >
          four
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-0-4"
          tabindex="-1"
        >
          five
        </td>
      </tr>
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="TableExample-tr-1"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-1-0"
          tabindex="-1"
        >
          one - 2
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-1-1"
          tabindex="-1"
        />
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-1-2"
          tabindex="-1"
        />
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-1-3"
          tabindex="-1"
        >
          four - 2
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-1-4"
          tabindex="-1"
        >
          five - 2
        </td>
      </tr>
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="TableExample-tr-2"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-2-0"
          tabindex="-1"
        >
          one - 3
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-2-1"
          tabindex="-1"
        >
          two - 3
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-2-2"
          tabindex="-1"
        >
          three - 3
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-2-3"
          tabindex="-1"
        >
          four - 3
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-2-4"
          tabindex="-1"
        >
          five - 3
        </td>
      </tr>
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="TableExample-tr-3"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-3-0"
          tabindex="-1"
        >
          one - 4
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-3-1"
          tabindex="-1"
        >
          two - 4
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-3-2"
          tabindex="-1"
        >
          null
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-3-3"
          tabindex="-1"
        >
          four - 4
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-3-4"
          tabindex="-1"
        >
          five - 4
        </td>
      </tr>
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="TableExample-tr-4"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-4-0"
          tabindex="-1"
        >
          one - 5
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-4-1"
          tabindex="-1"
        >
          two - 5
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-4-2"
          tabindex="-1"
        >
          three - 5
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-4-3"
          tabindex="-1"
        >
          four - 5
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-4-4"
          tabindex="-1"
        >
          five - 5
        </td>
      </tr>
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="TableExample-tr-5"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-5-0"
          tabindex="-1"
        >
          one - 6
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-5-1"
          tabindex="-1"
        >
          two - 6
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-5-2"
          tabindex="-1"
        >
          three - 6
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-5-3"
          tabindex="-1"
        >
          four - 6
        </td>
        <td
          class="pf-v5-c-table__td"
          data-ouia-component-id="TableExample-td-5-4"
          tabindex="-1"
        >
          five - 6
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
